src/types/shared/simulation-payload.ts

```diff
@@ -0,0 +1,102 @@
+  type SubscriptionImportedEvent,
+  type SubscriptionPastDueEvent,
+  type SubscriptionPausedEvent,
+  type SubscriptionResumedEvent,
+  type SubscriptionTrialingEvent,
+  type SubscriptionUpdatedEvent,
+  type TransactionBilledEvent,
+  type TransactionCanceledEvent,
+  type TransactionCompletedEvent,
+  type TransactionCreatedEvent,
+  type TransactionPaidEvent,
+  type TransactionPastDueEvent,
+  type TransactionPaymentFailedEvent,
+  type TransactionReadyEvent,
+  type TransactionUpdatedEvent,
+} from '../../notifications/events';
+
+interface SimulationEventPayloadMap {
+  'address.created': AddressCreatedEvent;
+  'address.updated': AddressUpdatedEvent;
+  'address.imported': AddressImportedEvent;
+  'adjustment.created': AdjustmentCreatedEvent;
+  'adjustment.updated': AdjustmentUpdatedEvent;
+  'business.created': BusinessCreatedEvent;
+  'business.updated': BusinessUpdatedEvent;
+  'business.imported': BusinessImportedEvent;
+  'customer.created': CustomerCreatedEvent;
+  'customer.updated': CustomerUpdatedEvent;
+  'customer.imported': CustomerImportedEvent;
+  'discount.created': DiscountCreatedEvent;
+  'discount.updated': DiscountUpdatedEvent;
+  'discount.imported': DiscountImportedEvent;
+  'payout.created': PayoutCreatedEvent;
+  'payout.paid': PayoutPaidEvent;
+  'price.created': PriceCreatedEvent;
+  'price.updated': PriceUpdatedEvent;
+  'price.imported': PriceImportedEvent;
+  'product.created': ProductCreatedEvent;
+  'product.updated': ProductUpdatedEvent;
+  'product.imported': ProductImportedEvent;
+  'subscription.created': SubscriptionCreatedEvent;
+  'subscription.past_due': SubscriptionPastDueEvent;
+  'subscription.activated': SubscriptionActivatedEvent;
+  'subscription.canceled': SubscriptionCanceledEvent;
+  'subscription.imported': SubscriptionImportedEvent;
+  'subscription.paused': SubscriptionPausedEvent;
+  'subscription.resumed': SubscriptionResumedEvent;
+  'subscription.trialing': SubscriptionTrialingEvent;
+  'subscription.updated': SubscriptionUpdatedEvent;
+  'transaction.billed': TransactionBilledEvent;
+  'transaction.canceled': TransactionCanceledEvent;
+  'transaction.completed': TransactionCompletedEvent;
+  'transaction.created': TransactionCreatedEvent;
+  'transaction.paid': TransactionPaidEvent;
+  'transaction.past_due': TransactionPastDueEvent;
+  'transaction.payment_failed': TransactionPaymentFailedEvent;
+  'transaction.ready': TransactionReadyEvent;
+  'transaction.updated': TransactionUpdatedEvent;
+  'report.created': ReportCreatedEvent;
+  'report.updated': ReportUpdatedEvent;
+  subscription_creation: never;
+  subscription_renewal: never;
+  subscription_pause: never;
+  subscription_resume: never;
+  subscription_cancellation: never;
+}
+
+export type DiscriminatedSimulationEventResponse<Base> = {
+  [K in keyof SimulationEventPayloadMap]: Base & {
+    type: K;
+    payload?: K extends IEventName ? Partial<SimulationEventPayloadMap[K]['data']> | null : null;
```

Comment:
Would it be possible to map directly to the notification entity? e.g. something like this:

```suggestion
    payload?: K extends IEventName ? Partial<SimulationEventPayloadMap[K]> | null : null;
```

e.g. `AddressNotification`
```ts
interface SimulationEventPayloadMap {
  'address.created': AddressNotification;
  'address.updated': AddressNotification;
  'address.imported': AddressNotification;
.......
```

Reply:
I had this conversation with Sam as well, essentially the event often changes the type i.e.
`public override readonly data: Omit<SubscriptionNotification, 'transactionId'>`
so it uses that instead of the raw Notification